### PR TITLE
scheduler: unify leaderSolver and peerSolver in hot read scheduler

### PR DESCRIPTION
**Pull request by lhy1024**
Signed-off-by: lhy1024 <admin@liudos.us>

### What problem does this PR solve?

Previously, we gave priority to the transfer leader and only used the move leader when there was no transfer leader, which is not necessarily the optimal solution, so this pr will be compared uniformly and then selected

### Check List

Tests

- Unit test

### Release note

```release-note
unify leaderSolver and peerSolver in hot read scheduler
```

### CODE REVIEW

File: server/schedulers/hot_region.go [hunk 1]
```diff
@@ -436,6 +457,13 @@ type solution struct {
 	progressiveRank int64
 }
 
+func (bs *balanceSolver) addPendingInfluence() bool {
```
Comment (Yisaer) [inline server/schedulers/hot_region.go:460]: The name is misleading I think.

**Review by Yisaer — COMMENTED** [review]
rest lgtm

File: server/schedulers/hot_region.go [hunk 2]
```diff
@@ -365,19 +365,37 @@ func (h *hotScheduler) addPendingInfluence(op *operator.Operator, srcStore, dstS
...
 	peerSolver := newBalanceSolver(h, cluster, read, movePeer)
-	ops = peerSolver.solve()
-	if len(ops) > 0 {
-		return ops
+	peerOps := peerSolver.solve()
+	if len(leaderOps) == 0 && len(peerOps) == 0 {
+		schedulerCounter.WithLabelValues(h.GetName(), "skip").Inc()
+		return nil
 	}
+	if len(leaderOps) == 0 && peerSolver.addPendingInfluence() {
```
Comment (rleungx) [inline server/schedulers/hot_region.go:376]: Previously, we prefer to balance by the leader. In this PR, it changes?

File: server/schedulers/hot_region_test.go [hunk 3]
```diff
@@ -891,12 +896,12 @@ func (s *testHotReadRegionSchedulerSuite) TestWithPendingInfluence(c *C) {
 			hb.(*hotScheduler).clearPendingInfluence()
 
 			op1 := hb.Schedule(tc)[0]
-			testutil.CheckTransferLeader(c, op1, operator.OpLeader, 1, 3)
-			// store byte/key rate (min, max): (6.6, 7.1) | 6.1 | (6, 6.5) | 5
+			testutil.CheckTransferPeer(c, op1, operator.OpHotRegion, 1, 4)
+			// store byte/key rate (min, max): (6.6, 7.1) | 6.1 | 6 | (5,5.5)
```
Comment (rleungx) [inline server/schedulers/hot_region_test.go:900]: Why it changes, can you explain it?

File: server/schedulers/hot_region.go (hunk 2 above)
Comment (lhy1024) [reply server/schedulers/hot_region.go:376]: yes, this pr will compared them and then selected

File: server/schedulers/hot_region_test.go (hunk 3 above)
Comment (lhy1024) [reply server/schedulers/hot_region_test.go:900]: In the past, by transfer leader from 1 to 3 and move peer from 1 to 4, we obtained a distribution of 6.1,6.1,6.5,5.5; with this pr, we achieve a better distribution by moving peer from 1 to 4 twice, i.e. 6.1.6.1,6,6

Comment (lhy1024) [reply server/schedulers/hot_region_test.go]: update

File: server/schedulers/hot_region.go (hunk 1 above)
Comment (Yisaer) [reply server/schedulers/hot_region.go:460]: what about `tryAddPendingInfluence` ?

**Comment by Yisaer** [comment]
cc @Yisaer

File: server/schedulers/hot_region.go (hunk 1 above)
Comment (Yisaer) [reply server/schedulers/hot_region.go:460]: use `addPendingInfluence` will be better

**Review by Yisaer — APPROVED** [review]
rest lgtm

File: server/schedulers/hot_region.go [hunk 4]
```diff
@@ -369,19 +369,37 @@ func (h *hotScheduler) addPendingInfluence(op *operator.Operator, srcStore, dstS
...
 	peerSolver := newBalanceSolver(h, cluster, read, movePeer)
-	ops = peerSolver.solve()
-	if len(ops) > 0 {
-		return ops
+	peerOps := peerSolver.solve()
+	if len(leaderOps) == 0 && len(peerOps) == 0 {
+		schedulerCounter.WithLabelValues(h.GetName(), "skip").Inc()
+		return nil
 	}
+	if len(leaderOps) == 0 && peerSolver.tryAddPendingInfluence() {
```
Comment (rleungx) [inline server/schedulers/hot_region.go:380]: If tryAddPendingInfluence is false, it should be returned directly?

File: server/schedulers/hot_region_test.go [hunk 5]
```diff
@@ -895,37 +900,40 @@ func (s *testHotReadRegionSchedulerSuite) TestWithPendingInfluence(c *C) {
...
 			op1 := hb.Schedule(tc)[0]
-			testutil.CheckTransferLeader(c, op1, operator.OpLeader, 1, 3)
-			// store byte/key rate (min, max): (6.6, 7.1) | 6.1 | (6, 6.5) | 5
+			testutil.CheckTransferPeer(c, op1, operator.OpHotRegion, 1, 4)
+			// After move-peer, store byte/key rate (min, max): (6.6, 7.1) | 6.1 | 6 | (5,5.5)
 
 			op2 := hb.Schedule(tc)[0]
 			testutil.CheckTransferPeer(c, op2, operator.OpHotRegion, 1, 4)
-			// store byte/key rate (min, max): (6.1, 7.1) | 6.1 | (6, 6.5) | (5, 5.5)
+			// After move-peer,  store byte/key rate (min, max): (6.1, 7.1) | 6.1 | 6 | (5, 6)
```
Comment (rleungx) [inline server/schedulers/hot_region_test.go:914]:
```suggestion
			// After move-peer, store byte/key rate (min, max): (6.1, 7.1) | 6.1 | 6 | (5, 6)
```

File: server/schedulers/hot_region_test.go [hunk 6]
```diff
@@ -895,37 +900,40 @@ func (s *testHotReadRegionSchedulerSuite) TestWithPendingInfluence(c *C) {
...
+		// Before schedule, store byte/key rate: 7.1 | 6.1 | 6 | 5
+		// Min and max from storeLoadPred. They will be generated in the comparison of current and future.
 		for i := 0; i < 20; i++ {
 			hb.(*hotScheduler).clearPendingInfluence()
 
 			op1 := hb.Schedule(tc)[0]
-			testutil.CheckTransferLeader(c, op1, operator.OpLeader, 1, 3)
-			// store byte/key rate (min, max): (6.6, 7.1) | 6.1 | (6, 6.5) | 5
+			testutil.CheckTransferPeer(c, op1, operator.OpHotRegion, 1, 4)
+			// After move-peer, store byte/key rate (min, max): (6.6, 7.1) | 6.1 | 6 | (5,5.5)
```
Comment (rleungx) [inline server/schedulers/hot_region_test.go:910]:
```suggestion
			// After move-peer, store byte/key rate (min, max): (6.6, 7.1) | 6.1 | 6 | (5, 5.5)
```

File: server/schedulers/hot_region.go [hunk 7]
```diff
@@ -369,19 +369,43 @@ func (h *hotScheduler) addPendingInfluence(op *operator.Operator, srcStore, dstS
...
+	peerOps := peerSolver.solve()
+	if len(leaderOps) == 0 && len(peerOps) == 0 {
+		schedulerCounter.WithLabelValues(h.GetName(), "skip").Inc()
+		return nil
+	}
+	if len(leaderOps) == 0 {
+		if peerSolver.tryAddPendingInfluence() {
+			return peerOps
+		}
+		return nil
```
Comment (rleungx) [inline server/schedulers/hot_region.go:384]: skip？

File: server/schedulers/hot_region.go [hunk 8]
```diff
@@ -393,15 +419,15 @@ func (h *hotScheduler) balanceHotWriteRegions(cluster opt.Cluster) []*operator.O
 	case s < int(schedulePeerPr*100):
 		peerSolver := newBalanceSolver(h, cluster, write, movePeer)
 		ops := peerSolver.solve()
-		if len(ops) > 0 {
+		if len(ops) > 0 && peerSolver.tryAddPendingInfluence() {
```
Comment (rleungx) [inline server/schedulers/hot_region.go:422]: Perhaps, we also need to do a comparison for write. The rest LGTM.

**Review by rleungx — APPROVED** [review]

**Review by HunDunDM — APPROVED** [review]

**Pull request merged**